File: packages/pvm-assemblyscript/assembly/host-functions/accumulate/solicit.ts
import { RESULT_CODE_PANIC } from '../../config'
import { getRequestValue, setRequestValue, decodeRequestTimeslots, encodeRequestTimeslots } from '../../codec'
import {
  ACCUMULATE_ERROR_FULL,
  ACCUMULATE_ERROR_HUH,
  AccumulateHostFunctionContext,
  BaseAccumulateHostFunction,
  HostFunctionResult,
} from './base'

// Deposit constants (Gray Paper)
const C_BASEDEPOSIT: u64 = u64(100) // Base deposit
const C_ITEMDEPOSIT: u64 = u64(10) // Per-item deposit
const C_BYTEDEPOSIT: u64 = u64(1) // Per-byte deposit

// Helper to check for u64 addition overflow
// Returns true if a + b would overflow
@inline
function wouldOverflowU64(a: u64, b: u64): bool {
  return a > u64.MAX_VALUE - b
}

/**
 * SOLICIT accumulation host function (Ω_S)
 *
 * Solicits preimage request
 *
 * Gray Paper Specification:
 * - Function ID: 23 (solicit)
 * - Gas Cost: 10
 * - Parameters: registers[7-8] = o, z
 *   - o: hash offset in memory (32 bytes)
 *   - z: size of the preimage
 * - Returns: registers[7] = OK or error code
 *
 * Gray Paper Logic:
 * 1. Read hash from memory (32 bytes)
 * 2. Check if request already exists:
 *    - If doesn't exist: create empty request []
 *    - If exists as [x, y]: append current timeslot to make [x, y, t]
 *    - Otherwise: error HUH
 * 3. Check if service has sufficient balance
 * 4. Update service account with new request
 */
export class SolicitHostFunction extends BaseAccumulateHostFunction {
  functionId: u64 = u64(23) // SOLICIT function ID (Gray Paper: 23)
  name: string = 'solicit'
  gasCost: u64 = u64(10)

  execute(context: AccumulateHostFunctionContext): HostFunctionResult {
    const registers = context.registers
    const ram = context.ram
    const implications = context.implications
    const timeslot = context.timeslot
    // Extract parameters from registers
    const hashOffset = u64(registers[7])
    const preimageLength = u64(registers[8])

    // Read hash from memory (32 bytes)
    // Gray Paper: If memory read fails (h = error), return PANIC without changing registers
    const readResult_hash = ram.readOctets(u32(hashOffset), u32(32))
    if (readResult_hash.faultAddress !== 0 || readResult_hash.data === null) {
      // Gray Paper line 911: (panic, registers_7, ...) when h = error
      // Do NOT modify registers - just return PANIC
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }
    const hashData = readResult_hash.data!

    // Get the current implications context
    const imX = implications.regular

    // Get current service account
    const accountEntry = this.findAccountEntry(imX.state.accounts, imX.id)
    if (accountEntry === null) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_HUH)
      return new HostFunctionResult(255) // continue execution
    }
    const serviceAccount = accountEntry.account

    // Look up existing request using rawCshKeyvals helper
    const existingRequestValue = getRequestValue(serviceAccount, u32(imX.id), hashData, preimageLength)

    // Determine new request state based on Gray Paper logic
    let newTimeslots: u32[]
    let isNewRequest = false

    if (existingRequestValue === null) {
      // Request doesn't exist - create empty request []
      newTimeslots = []
      isNewRequest = true
    } else {
      // Decode existing request
      const existingTimeslots = decodeRequestTimeslots(existingRequestValue)
      if (existingTimeslots === null) {
        this.setAccumulateError(registers, ACCUMULATE_ERROR_HUH)
        return new HostFunctionResult(255) // continue execution
      }

      if (existingTimeslots.length === 2) {
        // Request exists as [x, y] - append current timeslot to make [x, y, t]
        newTimeslots = [existingTimeslots[0], existingTimeslots[1], u32(timeslot)]
      } else {
        // Invalid request state - cannot solicit
        this.setAccumulateError(registers, ACCUMULATE_ERROR_HUH)
        return new HostFunctionResult(255) // continue execution
      }
    }

    // Calculate new items and octets if this is a new request
    // Gray Paper: items += 2 for each new request (h, z)
    // Gray Paper: octets += (81 + z) for each new request
    let newItems = serviceAccount.items
    let newOctets = serviceAccount.octets

    if (isNewRequest) {
      newItems = serviceAccount.items + u32(2)
      
      // Check for u64 overflow: octets + 81 + preimageLength
      // TypeScript uses bigint (no overflow), so we must detect overflow and return FULL
      const octetsIncrement = u64(81) + preimageLength
      if (wouldOverflowU64(u64(81), preimageLength) || 
          wouldOverflowU64(serviceAccount.octets, octetsIncrement)) {
        // Overflow would occur - in TypeScript this results in minBalance > balance
        this.setAccumulateError(registers, ACCUMULATE_ERROR_FULL)
        return new HostFunctionResult(255) // continue execution
      }
      newOctets = serviceAccount.octets + octetsIncrement
    }

    // Calculate new minimum balance
    // Gray Paper: a_minbalance = max(0, Cbasedeposit + Citemdeposit * a_items + Cbytedeposit * a_octets - a_gratis)
    const itemDeposit = C_ITEMDEPOSIT * u64(newItems)
    const byteDeposit = C_BYTEDEPOSIT * newOctets
    
    // Check for overflow in totalDeposit calculation
    if (wouldOverflowU64(C_BASEDEPOSIT, itemDeposit) ||
        wouldOverflowU64(C_BASEDEPOSIT + itemDeposit, byteDeposit)) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_FULL)
      return new HostFunctionResult(255) // continue execution
    }
    
    const totalDeposit = C_BASEDEPOSIT + itemDeposit + byteDeposit
    const newMinBalance = totalDeposit > serviceAccount.gratis ? totalDeposit - serviceAccount.gratis : u64(0)

    // Check if service has sufficient balance for the new request
    // Gray Paper: If newMinBalance > balance, return FULL error
    if (newMinBalance > serviceAccount.balance) {
      this.setAccumulateError(registers, ACCUMULATE_ERROR_FULL)
      return new HostFunctionResult(255) // continue execution
    }

    // Update the service account with the new request using rawCshKeyvals helper
    setRequestValue(serviceAccount, u32(imX.id), hashData, preimageLength, encodeRequestTimeslots(newTimeslots))

    // Update items and octets if this is a new request
    if (isNewRequest) {
      serviceAccount.items = newItems
      serviceAccount.octets = newOctets
    }

    // Set success result
    this.setAccumulateSuccess(registers)
    return new HostFunctionResult(255) // continue execution
  }
}

File: packages/pvm-rust/src/host_functions/accumulate/solicit.rs
//! SOLICIT accumulation host function (Ω_S). Gray Paper: function ID 23.
//! r7 = hash offset (o), r8 = preimage size (z). 1:1 with AS solicit.ts.
//! Read 32-byte hash; get/set request value; new request [] or append timeslot to [x,y]; FULL on overflow/insufficient balance.

use crate::config::{C_BASE_DEPOSIT, C_BYTE_DEPOSIT, C_ITEM_DEPOSIT, FUNC_SOLICIT};
use crate::host_functions::accumulate::base::{self, codes};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};
use crate::codec::{
    decode_request_timeslots, encode_request_timeslots, get_request_value, set_request_value,
};

const HASH_LEN: u32 = 32;
/// Gray Paper: octets increment per new request (81 + z).
const OCTETS_BASE_PER_REQUEST: u64 = 81;

pub struct SolicitHostFunction;

impl SolicitHostFunction {
    /// Returns true if a + b would overflow u64.
    fn would_overflow_u64(a: u64, b: u64) -> bool {
        a > u64::MAX - b
    }
}

impl HostFunction for SolicitHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_SOLICIT
    }
    fn name(&self) -> &'static str {
        "solicit"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let hash_offset = context.registers[7] as u32;
        let preimage_length = context.registers[8];

        // Read hash from memory (32 bytes). Gray Paper: panic when h = error; r7 unchanged.
        let read_result = context.ram.read_octets(hash_offset, HASH_LEN);
        if read_result.fault_address != 0 || read_result.data.is_none() {
            return HostFunctionResult::panic();
        }
        let hash_data = read_result.data.unwrap();
        if hash_data.len() != HASH_LEN as usize {
            return HostFunctionResult::panic();
        }

        // Current service account (imX). None → HUH.
        let service_account = match &mut context.service_account {
            Some(acc) => acc,
            None => {
                base::set_accumulate_error(context.registers, codes::HUH);
                return HostFunctionResult::continue_execution();
            }
        };
        let service_id = match context.service_id {
            Some(id) => id as u32,
            None => {
                base::set_accumulate_error(context.registers, codes::HUH);
                return HostFunctionResult::continue_execution();
            }
        };

        let existing_request_value = get_request_value(service_account, service_id, &hash_data, preimage_length);

        let (new_timeslots, is_new_request) = match existing_request_value {
            None => {
                // Request doesn't exist - create empty request []
                (vec![], true)
            }
            Some(ref value) => {
                let existing_timeslots = match decode_request_timeslots(value) {
                    Some(t) => t,
                    None => {
                        base::set_accumulate_error(context.registers, codes::HUH);
                        return HostFunctionResult::continue_execution();
                    }
                };
                if existing_timeslots.len() == 2 {
                    // [x, y] - append current timeslot to make [x, y, t]
                    let timeslot = match context.timeslot {
                        Some(t) => t as u32,
                        None => {
                            base::set_accumulate_error(context.registers, codes::HUH);
                            return HostFunctionResult::continue_execution();
                        }
                    };
                    let mut nt = existing_timeslots;
                    nt.push(timeslot);
                    (nt, false)
                } else {
                    base::set_accumulate_error(context.registers, codes::HUH);
                    return HostFunctionResult::continue_execution();
                }
            }
        };

        let (new_items, new_octets) = if is_new_request {
            let new_items = service_account.items + 2;
            let octets_increment = OCTETS_BASE_PER_REQUEST
                .checked_add(preimage_length)
                .filter(|_| !Self::would_overflow_u64(OCTETS_BASE_PER_REQUEST, preimage_length));
            let octets_increment = match octets_increment {
                Some(v) => v,
                None => {
                    base::set_accumulate_error(context.registers, codes::FULL);
                    return HostFunctionResult::continue_execution();
                }
            };
            if Self::would_overflow_u64(service_account.octets, octets_increment) {
                base::set_accumulate_error(context.registers, codes::FULL);
                return HostFunctionResult::continue_execution();
            }
            let new_octets = service_account.octets + octets_increment;
            (new_items, new_octets)
        } else {
            (service_account.items, service_account.octets)
        };

        // newMinBalance = max(0, totalDeposit - gratis). Check overflow in totalDeposit.
        let item_deposit = C_ITEM_DEPOSIT * new_items as u64;
        let byte_deposit = C_BYTE_DEPOSIT * new_octets;
        if Self::would_overflow_u64(C_BASE_DEPOSIT, item_deposit)
            || Self::would_overflow_u64(C_BASE_DEPOSIT + item_deposit, byte_deposit)
        {
            base::set_accumulate_error(context.registers, codes::FULL);
            return HostFunctionResult::continue_execution();
        }
        let total_deposit = C_BASE_DEPOSIT + item_deposit + byte_deposit;
        let new_min_balance = total_deposit.saturating_sub(service_account.gratis);

        if new_min_balance > service_account.balance {
            base::set_accumulate_error(context.registers, codes::FULL);
            return HostFunctionResult::continue_execution();
        }

        set_request_value(
            service_account,
            service_id,
            &hash_data,
            preimage_length,
            encode_request_timeslots(&new_timeslots),
        );

        if is_new_request {
            service_account.items = new_items;
            service_account.octets = new_octets;
        }

        base::set_accumulate_success(context.registers, codes::OK);
        HostFunctionResult::continue_execution()
    }
}
